designs/drift-hash-versioning.md

```diff
@@ -0,0 +1,162 @@
+# Karpenter Drift Hash Versioning with CRD
+
+## Summary of Drift Hashing Mechanism
+
+Drift utilizes a hashing mechanism to identify when a `NodeClaim` has been drifted from its owning `NodePool/EC2NodeClass`. Karpenter will hash fields that are within the `NodePool.Spec/EC2NodeClass.Spec`, excluding [special cases of drift](https://karpenter.sh/docs/concepts/disruption/#special-cases-on-drift) or [behavior fields](https://karpenter.sh/docs/concepts/disruption/#behavioral-fields). The `NodePool/EC2NodeClass` will be annotated with the hash. When a `NodeClaim` is created, the `NodePool` and `EC2NodeClass` hashes will be annotated on the `NodeClaim`. More details can be found in the [drift documentation](https://karpenter.sh/docs/concepts/disruption/#drift). 
+
+When a `NodePool/EC2NodeClass` is updated, a new hash will be annotated on the `NodePool/EC2NodeClass`. The `NodeClaim` disruption controller will periodically check if the `NodePool/EC2NodeClass` hash match to determine when a `NodeClaim` is drifted.
+
+## Background
+
+Karpenter's static drift mechanism works well on additive fields, but the hash will change for all nodes when a field is removed or a new default is set. Currently, the team is looking to set defaults for the fields kind and apiVersion on the nodeClassRef of the nodepool.spec.template.spec [#909](https://github.com/kubernetes-sigs/karpenter/issues/909). This will result in Karpenter drifting whole clusters as the `NodePool` CRD will be updated and the `NodePool/EC2NodeClass` hash will be re-calculated. The team is also blocked from fixing issue [#5447](https://github.com/aws/karpenter-provider-aws/issues/5447). The fix to the issue is adding volumeSize in the hashing mechanism of the `EC2NodeClass` (#5454)[https://github.com/aws/karpenter-provider-aws/pull/5454]. However, this will cause all `NodeClaims` to be drifted as soon as a customer upgrades their Karpenter CRDs in their cluster. These two issues raise the need for Karpenter to understand different versions of the `NodePool/EC2NodeClass` CRD during a drift evaluation.  The problem is that all nodes on a customer’s cluster will be drifted without any changes to the node configuration
+
+More generally, Karpenter will need to guard against the following cases during a Karpenter upgrade:
+
+1. A `NodePool/EC2NodeClass` adds default values to existing fields that are included in the hash calculations
```

Comment:
Slightly ambiguous here. If a new default is added and user previously didn't set an value, doesn't that mean the behavior is changed? Shouldn't it drift? Are there cases where new default values shouldn't be propagated?

Reply:
Unless this means that the value is set but underlying default is changed. In which case that should be explicitly mentioned to clarify

Reply:
The idea here is that if a karpenter CRD adds in new default value, it would be unexpected for customers to have all of their nodes be drifted, since they didn't make any changes to the resource itself. I would think customers would have an expectation that the drift takes place once they have made a changes to their resource not update the CRD scheme. Do you disagree that customers would find it unexpected to have all of their nodes drifted on a CRD change? I can make an update to make that point more clear

Reply:
I partially agree but at the same time have some concerns. Let's say you change the default value in an CRD update and you don't drift. Now users will have their resources "stuck" in an older CRD behavior for certain fields until they update. What if they never update? What if you continue to update the CRD? How many back versions do you need to track?

In addition, if I am this user and I read the docs to understand something and I read that the current version of the CRD default field to N and I know I deployed this version of the CRD shouldn't I expect N behavior? I didn't realize updating CRD doesn't change this behavior or I need to update my resource in order to pick up this change? What would I update if my resource works the way it is?

I guess this is where the ambiguous behavior shows up for me. I don't think I fully agree that when I update the CRD, I don't expect drift. I do actually because something changed however I agree that changing a CRD shouldn't cause my enter cluster to drift without control. 

I think what I need is clarity in behavior. 
- I don't want my default behavior to differ from expectation (matching the CRD defaults) just because I didn't update the resource.
- I should know when behavior will change during a CRD update. If you add a default I should be able to know ahead of time that updating my CRD will drift my cluster or a set of nodes.
- I need to control how drift happens if a majority of my nodes are going to drift because of a CRD update

Reply:
> I guess this is where the ambiguous behavior shows up for me. I don't think I fully agree that when I update the CRD, I don't expect drift. I do actually because something changed however I agree that changing a CRD shouldn't cause my entire cluster to drift without control.

> I need to control how drift happens if a majority of my nodes are going to drift because of a CRD update 

Trying to understand this point here. What do you mean by drift without control? Currently, customer have ways to control disruption. In the case of a CRD update, disruption will only occur to the extent a customer has allowed for it to happen. The main problem addressed is should karpenter consider all the nodes it has launched to be drifted due to a CRD change? Karpenter will not violate any of the disruption controls that customer provide

> I partially agree but at the same time have some concerns. Let's say you change the default value in an CRD update and you don't drift. Now users will have their resources "stuck" in an older CRD behavior for certain fields until they update. What if they never update their resource?

> I don't want my default behavior to differ from expectation (matching the CRD defaults) just because I didn't update the resource. My nodes should drift because there is a change in behavior defined in the CRD.

This is a good callout. We will need to make sure that customers are aware will need to roll their clusters to match the defaults. This would be through a docs update on the upgrade path. One example would be for them to expire their nodes. Customer should have control over that as opposed to Karpenter automatically drifting IMO. I will make sure to add that here.

> I should know when behavior will change before an CRD update. If you add a new default I should be able to know ahead of time that updating my CRD will drift my cluster or a set of nodes and change its behavior.

We will defiantly need to document to customer when a change that will cause `nodepool-hash` to be re-calculated and propagated down. This change will prevent the CRD update from drifting clusters.

Reply:
> Trying to understand this point here. What do you mean by drift without control? Currently, customer have ways to control disruption. In the case of a CRD update, disruption will only occur to the extent a customer has allowed for it to happen. The main problem addressed is should karpenter consider all the nodes it has launched to be drifted due to a CRD change? Karpenter will not violate any of the disruption controls that customer provide

True I think this is likely less of an issue.

> This is a good callout. We will need to make sure that customers are aware will need to roll their clusters to match the defaults. This would be through a docs update on the upgrade path. One example would be for them to expire their nodes. Customer should have control over that as opposed to Karpenter automatically drifting IMO. I will make sure to add that here.

Exactly. I guess the question I have here is we could recommend what users do but my only worry is will users know? Do users actively read their docs and are the docs consistently up to date? I mainly want to point out that from a user of Karpenter I don't always do these things and in my mind the ideal solution is one that assumes users don't read but yet provide a consistent and/or predictable behavior.